### PR TITLE
add option to indent marshalled JSON

### PR DESCRIPTION
**Pull request by ansel1**

### CODE REVIEW

**Comment by ansel1** [comment]
Not sure why this isn't triggering the travis build.

**Comment by dghubble** [comment]
Can you provide a rationale for the feature please? Indentation of POST'ed JSON seems intended to support servers written to depend on that indentation.

**Comment by ansel1** [comment]
No, not to support servers.  It's purely for debugging purposes, when logging requests and responses.  Is simplest to turn on the indenting at the point of marshaling.

File: sling.go [hunk 1]
```diff
@@ -207,6 +210,12 @@ func (s *Sling) BodyJSON(bodyJSON interface{}) *Sling {
 	return s
 }
 
+// IndentJSON sets whether BodyJSON indents the marshaled JSON
+func (s *Sling) IndentJSON(b bool) *Sling {
+	s.indentJSON = true
```
Comment (jspiro) [inline sling.go]: Shouldn't this be `= b` ?

**Comment by ansel1** [comment]
Fixed turning the flag off, and rebased onto new master

**Comment by dghubble** [comment]
hm, I'm not sure if I'd use this much. Why not have the custom logger handle this?

**Comment by ansel1** [comment]
I've accomplished this before via server middleware, but it's fiddly, because at that point, all you have is a mime-type and bytes.  You need to inspect the mime-type to decide whether the bytes actually are JSON, then attempt to indent the bytes and do something with the error if it doesn't work.  Then replace the Reader/ReadCloser in the request/response.  It's just simpler if you can indent the JSON when it's being created.  Most golang http server frameworks have the option to indent marshalled JSON, for this reason.  golang's own json package offers both functions to indent bytes, or to indent while marshaling/encoding.

**Comment by dghubble** [comment]
This could now be customized with a `BodyProvider` implementation.

Add a BodyProvider implementation which indents the body however you prefer,

```
type indentedJSONBodyProvider struct {
	payload interface{}
}

func (p indentedJSONBodyProvider) ContentType() string {
	return jsonContentType
}

func (p indentedJSONBodyProvider) Body() (io.Reader, error) {
	buf := &bytes.Buffer{}
	err := json.NewEncoder(buf).Encode(p.payload)
	if err != nil {
		return nil, err
	}
	dest := &bytes.Buffer{}
	json.Indent(dest, buf.Bytes(), "", "  ")
	return dest, nil
}
```

Set that as the `BodyProvider`.

```
bp := indentedJSONBodyProvider{payload: mystruct}
req, err := sling.New().Post(path).BodyProvider(bp).Request()
```

The builtin Body, BodyForm, and BodyJSON methods now make use of `BodyProvider` behind the scenes.

**Comment by ansel1** [comment]
This still feels quite awkward.  Typically, indentation is something you set globally at startup, based on a flag, like "debug" mode.  And it's pretty typically to want indentation turned on in development, testing, and staging, because it makes debugging so much easier.  

Because indentation can't be turned on via a property of the slinger, it's not a property which can propagate down to child slingers.  This would need code changes in almost all the endpoints.  

We can build our own build provider factory functions in whatever request/session/connection structs wrap the slinger, so we can globally switch which kind of JSON body provider we use everywhere, but it still seems unnecessarily fiddly, compared to a simple IndentJSON() method on Sling.  I get that it's an edge case, but it's a pretty minor API concession for the sake of convenience and ergonomics.

**Comment by jspiro** [comment]
I agree. Put differently, just because a feature isn't pure doesn't mean it
isn't practical. It's also quite minor.

I ended up forking sling and significantly modifying it. Sometimes that's
the open source way and sometimes opening PRs is.

In general Gophers would argue for fewer features, but I still feel the
proposal is excessively awkward given the common need.


From: Russ Egan <notifications@github.com> <notifications@github.com>
Reply: dghubble/sling
<reply@reply.github.com>
<reply@reply.github.com>
Date: October 11, 2016 at 6:14:26 AM
To: dghubble/sling <sling@noreply.github.com> <sling@noreply.github.com>
CC: Jono Spiro <github@lostinrecursion.net> <github@lostinrecursion.net>,
Comment <comment@noreply.github.com> <comment@noreply.github.com>
Subject:  Re: [dghubble/sling] add option to indent marshalled JSON (#22)

This still feels quite awkward. Typically, indentation is something you set

**Comment by dghubble** [comment]
Before, you'd either use Sling setters to encode data into `http.Request` Body or do it yourself by modifying the `http.Request`. Now you have the ability to define a `BodyProvider` that does that in a custom way if you like:

```
if debug {
    bp := indentedJSONBodyProvider{payload: mystruct}
    mysling.New().Post(path).BodyProvider(bp).Receive(blah)
} else {
    mysling.New().Post(path).BodyJSON(bp).Receive(blah)
}
```

> it's not a property which can propagate down to child slingers

The value itself propagates just like any other set value.

Its true, there is no indentation convenience method that propagates your desired indentation through all subsequent body encodings on child slings.

```
mysling.SetJSONIndent("", "    ")
```

Instead, you could have your body provider implementation respect a debug state you set to determine how the decoding is done.

**Pull request closed without merging**